manuals/regfexport.1
.Dd June 24, 2026
.Dt REGFEXPORT 1
.Os
.Sh NAME
.Nm regfexport
.Nd exports items stored in a Windows NT Registry File (REGF)
.Sh SYNOPSIS
.Nm regfexport
.Op Fl c Ar codepage
.Op Fl K Ar key_path
.Op Fl l Ar log_file
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm regfexport
is a utility to export items stored in a Windows NT Registry File (REGF)
.Pp
.Nm regfexport
is part of the
.Nm libregf
package.
.Nm libregf
is a library to access the Windows NT Registry File (REGF) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl K Ar key_path
show information about a specific key path
.It Fl l Ar log_file
logs information about the exported items
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# cregexport SYSTEM.DAT
cregexport 20260623
.sp
Key path:
Key:
.sp
Key path: SOFTWARE
Key: SOFTWARE
.sp
Key path: SOFTWARE\\Classes
Key: Classes
.sp
Key path: SOFTWARE\\Classes\\CLSID
Key: CLSID
Value: 0 (default)
Type: string (REG_SZ)
Data size: 0
Data:
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr regfinfo 1 ,
.Xr regfmount 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libregf/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* regfexport print when key is corrupted, like value
* when multi value is corrupted flag Registry value as corrupted

* change value item handle missing value name
* fix:
  - multiple successive open of transaction regf file
* clean up
  - move named key, value key etc into separate source files
  - libregf_hive_bins_list_read_element_data remove the need for + 4096
* dokan add classname and values support
* regfinfo:
  - do not print (default) for corrupt value, but something like (corrupt)
* libregf:
  - add file type definitions
  - add error tollerance for missing value key
  - add value key flags definitions function (+/-)
    check documentation for more values
  - make signal abort part of IO handle
* pyregf
  - as string function: if not a string try to represent the data as a string?
* profile large regf files
* libregf_value_item_values_get_data
  - improve data stream solution
* libregf_value_get_value_binary_data
  - optimize to read directly from block or stream
* get value: add support for an empty string ("") to retrieve the default value
* refactor libregf_value_item_values_read_element_data to hive bins list
* merge functionality of evttools registry_file with regftools
* code clean up
  - libregf_value_item_values implement libfdata_reference

Tools:
* regfinfo
  - print hive bins list size
  - print last part of filename stored in header?
  - size usage distribution
* regfmount:
  - fix list file of specific value success withouh (values)\
  - detect key names with (classname), (values)
  - detect value names with (default)
  - detect duplicate key and value names
  - fix regfmount_fuse_read need libregf_value_read_data_at_offset function ?
  - expose security descriptor
  - add mode to auto convert utf16 into utf8 ?
  - check access to (default) still works
  - deal with duplicate key/value names
  - test cloning (fix out of file handles issue)
* add manuals
  - regfexport.1
  - regfmount.1

* is the hive bins list placed correctly in the IO handle ?
  or should it be moved into file ? or other location ?
* in get key by path functions
  - handle root key name (ignore proto.hiv)
* add binary data value seek and read function ?
* replace libregf_hive_bin_get_cell_at_offset by get _data function

Debug output
* print file type description

Performance
* optimizing list of keys with a large amount of sub keys
  - use hash table/map ? search tree based on hash ?
* read the security key on demand ?

Library
* build registry key hierarchy
  - validate parent key offset
* string values
  - how are they stored ?
  - use codepage in string value conversions
* data recovery

Error tollerance:
* deal with cyclic keys, add max depth or track key offsets

20140929
* see `git log' for more recent change log
* removed README.macosx
* changes for project site move

20141009
* updated dependencies and corresponding changes

20141004
* update Python-bindings tests

20140905
* updated libfwsi version check

20140831
* bug fix in Python-bindings

20140827
* code clean

20140803
* worked on regfmount

20140801
* bug fix in Python-bindings
* worked on regfmount
  - moved values to sub directory to prevent value and key name clashes
  - added class name support

20140729
* updated dependencies

20140427
* fix for dealing with size mismatch between value and block list size
* improved string size corruption handling

20140323
* worked on Python bindings

20140321
* worked on Python bindings
* worked on setup.py

20140304
* worked on Python bindings

20140217
* worked on Python bindings

20140118
* bug fix for call to libfdata_tree_node_set_sub_nodes_data_range

20140112
* 2014 update
* updated dependencies

20131230
* worked on Python bindings

20131229
* updated dependencies
* worked on Dokan support for regfmount

20131013
* improved pyregf as integer functions

20131008
* improved pyregf as integer functions

20130929
* worked on setup.py, largely for MSI builds

20130922
* removed hardcoded codepages from value functions
* updated msvscpp files

20130916
* updated dependencies

20130901
* updated dependencies

20130830
* worked on automated tests

20130829
* fixed incorrect year in change log

20130821
* updated dependencies

20130819
* fix in pyregf value data as integer function for bounds check being too strict

20130724
* small changes

20130723
* worked on shell items support

20130720
* worked on shell items support

20130718
* updated dependencies
* removed unnecessary restriction in library include headers
* changes due to libfdata update

20130717
* worked on error tolerance
* removed LIBREGF_ITEM_FLAGS definition

20130716
* updated dependencies
* worked on tests
* bug fix for large values

20130609
* updated dependencies

20130608
* added support for REG_LINK as a string

20130604
* fixed multiple open issues

20130602
* updated dependencies

20130601
* pyregf: fix for retrieving default value by name

20130421
* worked on tests

20130420
* bug fix in dependencies
* changes for strings values with trailing data
* worked on tests

20130416
* small update for 64-bit build

20130414
* Textual changes
* updates in pyregf

20130413
* updated dependencies
* worked on tests
* fixed codepage 1255 restriction
* updates and bug fixes in pyregf

20130409
* fixed codepage 949 and 950 restriction

20130405
* updated dependencies
* changes for libfdata update

20130402
* updated dependencies
* changes for libfdata update
* removed item flags

20130319
* updated dependencies
* changes for libfdata update

20130307
* worked on libfdata update
* updated macosx files

20130306
* worked on libfdata update

20130303
* added macosx package files
* changes for libfdata update

20130113
* updated msvscpp files

20130110
* worked on pyregf
* API changes to no longer expose internal data pointers
* worked on registry_file merge

20130108
* 2013 update
* updated dependencies
* worked on pyregf

20121127
* regfmount: worked on exposing class name

20121121
* pyregf: code clean up

20121112

regftools/regfexport.c
/*
 * Exports information from a Windows NT Registry File (REGF)
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "export_handle.h"
#include "regftools_getopt.h"
#include "regftools_libcerror.h"
#include "regftools_libclocale.h"
#include "regftools_libcnotify.h"
#include "regftools_libregf.h"
#include "regftools_output.h"
#include "regftools_signal.h"
#include "regftools_unused.h"

export_handle_t *regfexport_export_handle = NULL;
int regfexport_abort                      = 0;

/* Signal handler for regfexport
 */
void regfexport_signal_handler(
      regftools_signal_t signal REGFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "regfexport_signal_handler";

	REGFTOOLS_UNREFERENCED_PARAMETER( signal )

	regfexport_abort = 1;

	if( regfexport_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     regfexport_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use regfexport to export information from a Windows NT Registry File (REGF).";

	regftools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'K', "key_path", "show information about a specific key path" },
		{ 'l', "log_file", "logs information about the exported items" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error            = NULL;
	log_handle_t *log_handle            = NULL;
	system_character_t *option_codepage = NULL;
	system_character_t *key_path        = NULL;
	system_character_t *log_filename    = NULL;
	system_character_t *source          = NULL;
	char *program                       = "regfexport";
	system_integer_t option             = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( regftools_option_t ) );
	int result                          = 0;
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "regftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( regftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	regftools_output_version_fprint(
	 stdout,
	 program );

	if( regftools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = regftools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				regftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				regftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'K':
				key_path = optarg;

				break;

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				regftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		regftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libregf_notify_set_stream(
	 stderr,
	 NULL );
	libregf_notify_set_verbose(
	 verbose );

	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize log handle.\n" );

		goto on_error;
	}
	if( export_handle_initialize(
	     &regfexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize export handle.\n" );

		goto on_error;
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	if( option_codepage != NULL )
	{
		result = export_handle_set_ascii_codepage(
		          regfexport_export_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in export handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( export_handle_open_input(
	     regfexport_export_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( key_path != NULL )
	{
		if( export_handle_export_key_path(
		     regfexport_export_handle,
		     key_path,
		     log_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to export key path.\n" );

			goto on_error;
		}
	}
	else
	{
		if( export_handle_export_file(
		     regfexport_export_handle,
		     log_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to export file.\n" );

			goto on_error;
		}
	}
	if( export_handle_close_input(
	     regfexport_export_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close export handle.\n" );

		goto on_error;
	}
	if( export_handle_free(
	     &regfexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log handle.\n" );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( regfexport_export_handle != NULL )
	{
		export_handle_free(
		 &regfexport_export_handle,
		 NULL );
	}
	if( log_handle != NULL )
	{
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

